### ISSUE TITLE
Possible Bug in Mosaic Probability

### ISSUE THREAD

**Issue by jaideep11061982** [ultralytics/yolov5#6477]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training

### Bug

I am passing mosaci in hyp.scratch.yaml as 1. Which means mosaci should always work . But I find that i isnt find it to be case as per below debug statements i put it in Datasets LoadImageLabels  classes' Init and get item class method

```
self.mosaic init True augment True rect False  --as per scratch hyp.yaml  for Train
train: Scanning '/kaggle/working/COTS/labels/train.cache' images and labels... 4
get item hyp mosaic 1.0 True self.mosaic True
get item hyp mosaic 1.0 True self.mosaic True
val loader
self.mosaic init False augment False rect True  --for val we dont enable mosaic  as expected.
val: Scanning '/kaggle/working/COTS/labels/valid.cache' images and labels... 577
get item hyp mosaic 1.0 False self.mosaic False  - Training begins . Mysteriously self.mosaic is false now . 
```
<this debug is inside this
```
   def __getitem__(self, index):
        index = self.indices[index]  # linear, shuffled, or image_weights

        hyp = self.hyp
        mosaic = self.mosaic and random.random() < hyp['mosaic']
        print('get item hyp mosaic',hyp['mosaic'],mosaic,'self.mosaic',self.mosaic)
```

get item hyp mosaic 1.0 False self.mosaic False

### Minimal Reproducible Example

https://github.com/ultralytics/yolov5/blob/7539cd75c3a6c06d00848617f6265f39a765ccea/utils/datasets.py

```
class LoadImagesAndLabels(Dataset):
    # YOLOv5 train_loader/val_loader, loads images and labels for training and validation
    cache_version = 0.6  # dataset labels *.cache version

    def __init__(self, path, img_size=640, batch_size=16, augment=False, hyp=None, rect=False, 
                 image_weights=False,
                 cache_images=False, single_cls=False, stride=32, pad=0.0, prefix=''):
        self.img_size = img_size
        self.augment = augment
        self.hyp = hyp
        self.image_weights = image_weights
        self.rect = False if image_weights else rect
        self.mosaic = self.augment and not self.rect  # load 4 images at a time into a mosaic (only during training)
        print('self.mosaic init',self.mosaic,'augment',self.augment,'rect',self.rect)
```

```
def __getitem__(self, index):
        index = self.indices[index]  # linear, shuffled, or image_weights

        hyp = self.hyp
        mosaic = self.mosaic and random.random() < hyp['mosaic']+1
        print('get item hyp mosaic',hyp['mosaic'],mosaic,'self.mosaic',self.mosaic)
```

### Additional
Moreoever i also see that whenever I set mosaic as complete 0 in hyp.yaml file, letter box  gives this wierd image resize
```
load image rect (1800, 3200, 3)   720 1280 change 1800 3200
letterbox (3200, 3200, 3)
```
Above is inside this

```
class LoadImagesAndLabels(Dataset):
......
.......
def __getitem__(self, index):
        index = self.indices[index]  # linear, shuffled, or image_weights

        hyp = self.hyp
        p=random.random()
        mosaic = self.mosaic and p < hyp['mosaic'] 
        print('get item hyp mosaic',hyp['mosaic'],mosaic,'self.mosaic',self.mosaic,p)
        if mosaic:
            # Load mosaic
            img, labels = load_mosaic(self, index)
            print('mosaic',img.shape)
            shapes = None

            # MixUp augmentation
            if random.random() < hyp['mixup']:
                img, labels = mixup(img, labels, *load_mosaic(self, random.randint(0, self.n - 1)))

        else:
            # Load image
            img, (h0, w0), (h, w) = load_image(self, index)
            print('load image rect',img.shape,' ',h0,w0 ,'change',h,w)
            # Letterbox
            shape = self.batch_shapes[self.batch[index]] if self.rect else self.img_size  # final letterboxed shape
            img, ratio, pad = letterbox(img, shape, auto=False, scaleup=self.augment)
            print('letterbox',img.shape)
```
_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@jaideep11061982 everything is working correctly. Mosaic is False during validation.

**Comment by jaideep11061982** [comment]
Hi @glenn-jocher  
Issue is not related to Validation but the training. 
If you see debug statements. You can see that during training self.mosaic is getting to False/True  even though it gets set to True during Training dataloader initialization as per Hyp mosaic setting. I coudlnt find any code which is randomly setting it to False during Training Epoch

**Comment by jaideep11061982** [comment]
See more trail of logs

```
load image rect (1800, 3200, 3)   720 1280 change 1800 3200
letterbox (1856, 3264, 3)
load image rect (1800, 3200, 3)   720 1280 change 1800 3200
letterbox (1856, 3264, 3)
get item hyp mosaic 1.0 False self.mosaic False
get item hyp mosaic 1.0 False self.mosaic False
get item hyp mosaic 1.0 True self.mosaic True
load image rect (1800, 3200, 3)   720 1280 change 1800 3200
get item hyp mosaic 1.0 True self.mosaic True
```
Self.mosaic is getting Flip/Flop  during batch Load.

**Comment by jaideep11061982** [comment]
my understanding is we should never go else of if mosaic  during the training as we pass mosaci=1.0 in yaml file.

**Comment by glenn-jocher** [comment]
@jaideep11061982 your console output shows this happening during val obviously, i.e. 

```python
val: Scanning '/kaggle/working/COTS/labels/valid.cache' images and labels... 577
get item hyp mosaic 1.0 False self.mosaic False  - Training begins . Mysteriously self.mosaic is false now . 
```

**Comment by jaideep11061982** [comment]
@glenn-jocher 
val.py is called after an epoch. Only first time when we init Val loader we will see val specific mosaic/rec value as you see i have put an indicator just before we create validation loader,once that is done we now iterate through Training images in train.py.  I see no code that is parallelly also fetching validation items. 

**val loader
self.mosaic init False augment False rect True  --for val we dont enable mosaic  as expected.**
val: Scanning '/kaggle/working/COTS/labels/valid.cache' images and labels... 577


from train.py

```
mloss = torch.zeros(3, device=device)  # mean losses
 if RANK != -1:
            train_loader.sampler.set_epoch(epoch)
 pbar = enumerate(train_loader)
 LOGGER.info(('\n' + '%10s' * 7) % ('Epoch', 'gpu_mem', 'box', 'obj', 'cls', 'labels', 'img_size'))
  if RANK in [-1, 0]:
            pbar = tqdm(pbar, total=nb, bar_format='{l_bar}{bar:10}{r_bar}{bar:-10b}')  # progress bar
optimizer.zero_grad()
if RANK in [-1, 0]:
            pbar = tqdm(pbar, total=nb, bar_format='{l_bar}{bar:10}{r_bar}{bar:-10b}')  # progress bar
        optimizer.zero_grad()
        for i, (imgs, targets, paths, _) in pbar:  # batch -------------------------------------------------------------
            ni = i + nb * epoch  # number integrated batches (since train start)
            imgs = imgs.to(device, non_blocking=True).float() / 255  # uint8 to float32, 0-255 to 0.0-1.0

            # Warmup
            if ni <= nw:
                xi = [0, nw]  # x interp
                # compute_loss.gr = np.interp(ni, xi, [0.0, 1.0])  # iou loss ratio (obj_loss = 1.0 or iou)
                accumulate = max(1, np.interp(ni, xi, [1, nbs / batch_size]).round())
                for j, x in enumerate(optimizer.param_groups):
                    # bias lr falls from 0.1 to lr0, all other lrs rise from 0.0 to lr0
                    x['lr'] = np.interp(ni, xi, [hyp['warmup_bias_lr'] if j == 2 else 0.0, x['initial_lr'] * lf(epoch)])
                    if 'momentum' in x:
                        x['momentum'] = np.interp(ni, xi, [hyp['warmup_momentum'], hyp['momentum']])

            # Multi-scale
            if opt.multi_scale:
                #sz = random.randrange(imgsz * 0.5, imgsz * 1.5 + gs) // gs * gs  # size
                sz = random.randrange(imgsz * 0.7, imgsz * 2 + gs) // gs * gs  # size

                sf = sz / max(imgs.shape[2:])  # scale factor
                if sf != 1:
                    ns = [math.ceil(x * sf / gs) * gs for x in imgs.shape[2:]]  # new shape (stretched to gs-multiple)
                    imgs = nn.functional.interpolate(imgs, size=ns, mode='bilinear', align_corners=False)

            # Forward
            with amp.autocast(enabled=cuda):
                pred = model(imgs)  # forward
                loss, loss_items = compute_loss(pred, targets.to(device))  # loss scaled by batch_size
                if RANK != -1:
                    loss *= WORLD_SIZE  # gradient averaged between devices in DDP mode
                if opt.quad:
                    loss *= 4.

            # Backward
            scaler.scale(loss).backward()

            # Optimize
            if ni - last_opt_step >= accumulate:
                scaler.step(optimizer)  # optimizer.step
                scaler.update()
                optimizer.zero_grad()
                if ema:
                    ema.update(model)
                last_opt_step = ni

            # Log
            if RANK in [-1, 0]:
                mloss = (mloss * i + loss_items) / (i + 1)  # update mean losses
                mem = f'{torch.cuda.memory_reserved() / 1E9 if torch.cuda.is_available() else 0:.3g}G'  # (GB)
                pbar.set_description(('%10s' * 2 + '%10.4g' * 5) % (
                    f'{epoch}/{epochs - 1}', mem, *mloss, targets.shape[0], imgs.shape[-1]))
                callbacks.run('on_train_batch_end', ni, model, imgs, targets, paths, plots, opt.sync_bn)
            # end batch ------------------------------------------------------------------------------------------------

        # Scheduler
        lr = [x['lr'] for x in optimizer.param_groups]  # for loggers
        scheduler.step()

        if RANK in [-1, 0]:
            # mAP
            callbacks.run('on_train_epoch_end', epoch=epoch)
            ema.update_attr(model, include=['yaml', 'nc', 'hyp', 'names', 'stride', 'class_weights'])
            final_epoch = (epoch + 1 == epochs) or stopper.possible_stop
            if not noval or final_epoch:  # Calculate mAP
                results, maps, _,f2 = val.run(data_dict,
```

**Comment by jaideep11061982** [comment]
ok @glenn-jocher  i put val/train flag and found that only flag for val is on mosaic is false. 
so I am fine with this.

But one more thing if you can help and then please close the bug.

is  when mosaic is why does letter box  function  gives sz as   max(h,w) * max(h,w) ..
